### PR TITLE
Noise gate demo integrated with the worklet

### PR DESCRIPTION
**Pull request by samhincks**
In this PR, audio-worklet/noisegate.html and audio-worklet/src/noisegate-demo.js have been modified to support audio-worklet/src/noisegate-audio-worklet.js. The audio worklet version only works if the browser supports audio worklets. 

The demo can be found [here](http://rawgit.com/GoogleChrome/web-audio-samples/2d62ef30ef41c7850c394428767c0e53a68ee695/audio-worklet/noisegate.html).

### CODE REVIEW

File: audio-worklet/noisegate.html [hunk 1]
```diff
@@ -54,20 +55,33 @@
...
-      beginNoisegateDemo();
+      // If the browser implements AudioWorklet, the demo enables the user to
+      // switch between audio worklet and script processor implementations of
+      // the noise gate.
+      const workletIsAvailable = AudioWorkletHelper.isAvailable();
       
+      if (workletIsAvailable) {
+        window.audioWorklet.addModule('src/noisegate-audio-worklet.js')
+            .then(() => {
+                beginNoisegateDemo();
```
Comment (hoch) [inline audio-worklet/noisegate.html]: `then(beginNodegateDemo);` is good enough.

File: audio-worklet/src/noisegate-demo.js [hunk 2]
```diff
@@ -36,12 +35,33 @@ class NoisegateDemo {
...
-    // TODO: Implement audio worklet logic.
+    
+    if (this.workletIsAvailable_) {
+      this.noisegateAudioWorklet_ =
+          new AudioWorkletNode(this.context_, 'noisegate-audio-worklet');
+
+      // The script processor is used by default, and if workletGain_.gain is 0,
+      // then scriptProcessorGain_.gain is 1, and vice versa.
+      this.workletGain_ = new GainNode(this.context_, {gain: 0});
+      this.speechAndNoiseSummingJunction_.connect(
```
Comment (hoch) [inline audio-worklet/src/noisegate-demo.js]: Is this clang-format? Really weird.

File: audio-worklet/src/noisegate-demo.js [hunk 3]
```diff
@@ -67,6 +80,7 @@ class NoisegateDemo {
     this.speechAndNoiseSummingJunction_.connect(
         this.noisegateScriptProcessor_.input);
     this.noisegateScriptProcessor_.output.connect(
```
Comment (hoch) [inline audio-worklet/src/noisegate-demo.js]: Yep. I speculate clang-format is doing this...

File: audio-worklet/src/noisegate-demo.js [hunk 4]
```diff
@@ -187,6 +206,19 @@ class NoisegateDemo {
     }
   }
 
+  workletSelected() {
+    // Switching occurs 10ms into the future for thread synchronization.
```
Comment (hoch) [inline audio-worklet/src/noisegate-demo.js]: What does this mean? Why do you need 0.01s?

Comment (samhincks) [inline audio-worklet/src/noisegate-demo.js]: We scheduled switching between script processor and audio worklet implementations into the future in the bitcrusher for reasons having to do with that the script processor operates in a separate thread and at a slight delay, but I suppose I don't fully understand the mechanics of why this is necessary. I defer to @rtoy that the comment and scheduling is correct.

File: audio-worklet/src/noisegate-demo.js (hunk 2 above)
Comment (samhincks) [inline audio-worklet/src/noisegate-demo.js]: It got reformatted with Clang. I have made sure the rest of the document is consistent with Clang now.

**Comment by samhincks** [comment]
In the latest [commit](https://github.com/GoogleChrome/web-audio-samples/pull/114/commits/7b60d65641fbae0ffe11d049679c4f1548c63b09),
- modifications based on review.

File: audio-worklet/noisegate.html [hunk 5]
```diff
@@ -54,20 +55,31 @@
...
-      const workletIsAvailable = false;
-      beginNoisegateDemo();
+      // If the browser implements AudioWorklet, the demo enables the user to
+      // switch between audio worklet and script processor implementations of
+      // the noise gate.
+      const workletIsAvailable = AudioWorkletHelper.isAvailable();
       
+      if (workletIsAvailable) {
+        window.audioWorklet.addModule('src/noisegate-audio-worklet.js')
+            .then(beginNoisegateDemo());
```
Comment (rtoy) [inline audio-worklet/noisegate.html]: I think this is wrong.

Comment (hoch) [inline audio-worklet/noisegate.html]: You should do `.then(beginNoisegateDemo);`

File: audio-worklet/src/noisegate-demo.js (hunk 4 above)
Comment (hoch) [inline audio-worklet/src/noisegate-demo.js]: If you want to take account of different latency between SPN and AWN, the value should not be `10ms`. I suggest to remove 0.01s here and simply use the current time for both. I don't expect A-Bing to be completely glitchless.

File: audio-worklet/src/noisegate-demo.js [hunk 6]
```diff
@@ -244,12 +273,13 @@ class NoisegateDemo {
...
-      this.releaseSlider_.disable();
-      this.thresholdSlider_.disable();
-    }
+    this.speechSource_.onended =
+        () => {
+          this.sourceButton_.enable();
+          this.attackSlider_.disable();
+          this.releaseSlider_.disable();
+          this.thresholdSlider_.disable();
+        }
```
Comment (hoch) [inline audio-worklet/src/noisegate-demo.js]: The JS linter should have given you a warning for a missing semicolon here.

**Comment by samhincks** [comment]
In the latest [commit](https://github.com/GoogleChrome/web-audio-samples/pull/114/commits/4f43707a4b9137882f1681ff2db3a4ac98f7071e),
- modifications based on review.

**Pull request merged**